plugins/formal-verification/formver.uniqueness/src/org/jetbrains/kotlin/formver/UniqueCFA.kt

```diff
@@ -0,0 +1,171 @@
+
+class UniqueCFA(private val data: UniqueCheckerContext) : FirControlFlowChecker(MppCheckerKind.Common) {
+    override fun analyze(graph: ControlFlowGraph, reporter: DiagnosticReporter, context: CheckerContext) {
+        graph.traverseToFixedPoint(UniqueInfoCollector(this.data))
+    }
+
+    private class UniqueInfoCollector(private val context: UniqueCheckerContext) :
+        PathAwareControlFlowGraphVisitor<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>() {
+
+        override fun mergeInfo(
+            a: ControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+            b: ControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+            node: CFGNode<*>,
+        ): ControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            // TODO: No implemented yet, for now just return from one branch
+            return a
+        }
+
+        override fun visitEdge(
+            from: CFGNode<*>,
+            to: CFGNode<*>,
+            metadata: Edge,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForEdge = super.visitEdge(from, to, metadata, data)
+            return dataForEdge
+        }
+
+        override fun visitNode(
+            node: CFGNode<*>,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            if (node is EnterNodeMarker) {
+                context.uniqueStack.add(ArrayDeque())
+            } else if (node is ExitNodeMarker) {
+                context.uniqueStack.popLast().lastOrNull()?.let { context.uniqueStack.last().add(it) }
+            }
+            return data
+        }
+
+        override fun visitFunctionEnterNode(
+            node: FunctionEnterNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForNode = visitNode(node, data)
+            val declaration = node.fir
+
+            // construct context and start control flow analysis
+            if (declaration.receiverParameter != null) {
+                TODO("Ignore receiver for now, not sure how to convert it into property symbol yet")
+            }
+            val valueParameters =
+                declaration.valueParameters.associate {
+                    Pair<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>(
+                        it.symbol,
+                        if (it.hasAnnotation(this.context.uniqueId, this.context.session)) {
+                            setOf(UniqueLevel.Unique)
+                        } else {
+                            setOf(UniqueLevel.Shared)
+                        }
```

Comment:
This sounds like something you might want to do often; please split it to a helper function.

Reply:
This is done in the other draft pr, I will bring them here